samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -173,71 +226,151 @@ public void deleteStartpoint(SystemStreamPartition ssp, TaskName taskName) {
+        }
+        if (deleteKeys.remove(ssp, null)) {
+          deleteStartpoint(ssp);
         }
       }
     }
 
-    // Delete SSP-only keys
-    sspsToDelete.forEach(ssp -> {
-        deleteStartpoint(ssp);
-        LOG.info("All Startpoints for SSP: {} have been grouped to the appropriate tasks and the SSP was deleted.");
-      });
+    return ImmutableSet.copyOf(fanOuts.keySet());
+  }
+
+  /**
+   * Read the fanned out {@link Startpoint}s for the given {@link TaskName}
+   * @param taskName to read the fan out Startpoints for
+   * @return fanned out Startpoints
+   */
+  public Map<SystemStreamPartition, Startpoint> getFanOutForTask(TaskName taskName) {
+    Preconditions.checkState(!stopped, "Underlying metadata store not available");
+    Preconditions.checkNotNull(taskName, "TaskName cannot be null");
 
-    return ImmutableSet.copyOf(sspsToDelete);
+    StartpointFanOut.StartpointFanOutSerde startpointFanOutSerde = new StartpointFanOut.StartpointFanOutSerde();
+    byte[] fanOutBytes = fanOutStore.get(toFanOutStoreKey(taskName));
+    if (ArrayUtils.isEmpty(fanOutBytes)) {
+      return null;
+    }
+    StartpointFanOut startpointFanOut = startpointFanOutSerde.fromBytes(fanOutBytes);
+    return ImmutableMap.copyOf(startpointFanOut.getFanOuts());
   }
 
   /**
-   * Relinquish resources held by the underlying {@link MetadataStore}
+   * Deletes the fanned out {@link Startpoint} for the given {@link TaskName}
+   * @param taskName to delete the fan out Startpoints for
    */
-  public void stop() {
-    stopped = true;
-    // Metadata store lifecycle is managed outside of the StartpointManager, so not closing it.
+  public void removeFanOutForTask(TaskName taskName) {
+    Preconditions.checkState(!stopped, "Underlying metadata store not available");
+    Preconditions.checkNotNull(taskName, "TaskName cannot be null");
+
+    fanOutStore.delete(toFanOutStoreKey(taskName));
   }
 
   @VisibleForTesting
   MetadataStore getMetadataStore() {
     return metadataStore;
   }
 
-  private static String toStoreKey(SystemStreamPartition ssp, TaskName taskName) {
+  @VisibleForTesting
+  MetadataStore getReadWriteStore() {
+    return readWriteStore;
+  }
+
+  @VisibleForTesting
+  MetadataStore getFanOutStore() {
+    return fanOutStore;
+  }
+
+  private static Startpoint resolveStartpointPrecendence(Startpoint startpoint1, Startpoint startpoint2) {
+    if (startpoint1 != null && startpoint2 != null) {
+      // if SSP-only and SSP+taskName startpoints both exist, resolve to the one with the latest timestamp
+      if (startpoint1.getCreationTimestamp() > startpoint2.getCreationTimestamp()) {
+        return startpoint1;
+      }
+      return startpoint2;
+    }
+    return startpoint1 != null ? startpoint1 : startpoint2;
+  }
+
+  private static String toReadWriteStoreKey(SystemStreamPartition ssp, TaskName taskName) {
     return new String(new JsonSerdeV2<>().toBytes(new StartpointKey(ssp, taskName)));
```

Comment:
Is `JsonSerdeV2` guaranteed to keep the same ordering of keys across different JVMs?
For example, {"ssp": "ssp_value", "taskName": "taskName_value"} is the same semantic key as {"taskName": "taskName_value", "ssp": "ssp_value"}, but the byte representation is different.

Reply:
Yes if the keys are json object itself are the same. However, makes sense to put stronger guarantees. I'll do that.